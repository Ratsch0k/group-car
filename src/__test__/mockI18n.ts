jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
}));
