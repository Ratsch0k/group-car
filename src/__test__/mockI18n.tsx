import React from 'react';

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (str: string) => str,
      i18n: {
        changeLanguage: () => new Promise(() => undefined),
        languages: ['de', 'en'],
        language: 'en',
      },
    };
  },
  Trans: (props: any) => {
    return <div>{props.i18nKey}</div>;
  },
}));
