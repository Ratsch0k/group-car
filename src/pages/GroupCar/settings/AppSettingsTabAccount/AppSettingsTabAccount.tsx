import React, {FC} from 'react';
import {
  Box,
} from '@material-ui/core';
import {useTranslation} from 'react-i18next';
import {UserInfo} from './UserInfo';
import SettingsSection from 'lib/components/SettingsSection';
import Security from './Security';
import SettingsTabTitle from 'lib/components/Settings/SettingsTabTitle';


export const AppSettingsTabAccount: FC = () => {
  const {t} = useTranslation();

  return (
    <>
      <SettingsTabTitle>
        {t('settings.account.title')}
      </SettingsTabTitle>
      <Box>
        <SettingsSection title={t('settings.account.userInfo')}>
          <UserInfo />
        </SettingsSection>
        <SettingsSection title={t('settings.account.security.title')}>
          <Security />
        </SettingsSection>
      </Box>
    </>
  );
};

export default AppSettingsTabAccount;
