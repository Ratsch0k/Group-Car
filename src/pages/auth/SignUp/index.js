export {default as SignUpForm} from './SignUpForm';
export {default as SignUpDialog} from './SignUpDialog';
