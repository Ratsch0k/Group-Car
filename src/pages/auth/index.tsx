export {default} from './Authentication';
export * from './Login';
export * from './SignUp';
