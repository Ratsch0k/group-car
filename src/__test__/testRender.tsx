/* eslint-disable @typescript-eslint/no-explicit-any */
import './mockAxios';
import './mockI18n';
import React from 'react';
import {render, RenderResult} from '@testing-library/react';
import {ConnectedRouter} from 'connected-react-router';
import {Provider} from 'react-redux';
import history from '../lib/redux/history';
import {StylesProvider} from '@material-ui/styles';
import {MuiThemeProvider} from '@material-ui/core';
import testTheme from './testTheme';
import {
  MockStoreEnhanced,
} from 'redux-mock-store';
import {RootState} from '../lib/redux/store';
import mockStore from './mockStore';
import mediaQuery from 'css-mediaquery';

export interface GlobalMocks {
  ResizeObserver: jest.Mock<{observe: jest.Mock, disconnect: jest.Mock}>;
}


export interface TestRenderResult {
  store: MockStoreEnhanced<unknown>;
  globalMocks: GlobalMocks;
}

/**
 * Testing generateClassName function to
 * avoid weird fails due to random class names
 * @param rule Rule
 * @param stylesheet Stylesheet
 * @returns The class name
 */
const generateClassName = (rule: any, stylesheet: any) => {
  return `${stylesheet.options.classNamePrefix}-${rule.key}`;
};

export interface TestRenderOptions {
  width?: string;
}

/**
 * Creates a match media to fix useMediaQuery in tests.
 * @param width Width of window
 * @returns Match media mock
 */
function createMatchMedia(width: any) {
  return (query: any) => ({
    matches: mediaQuery.match(query, {width}),
    addListener: () => undefined,
    removeEventListener: () => undefined,
  });
}


/**
 * Mock all global function which are used.
 */
const mockGlobals = (width: number): GlobalMocks => {
  // Mock ResizeObserver
  const resizeObserverMock = jest.fn().mockImplementation(() => ({
    observe: jest.fn(),
    disconnect: jest.fn(),
  }));
  window.ResizeObserver = resizeObserverMock;
  window.innerWidth = width;
  window.matchMedia = createMatchMedia(window.innerWidth) as any;

  return {ResizeObserver: resizeObserverMock};
};

/**
 * Render function for testing purposes.
 * Provides the necessary base components and
 * contexts.
 * @param state Redux state
 * @param children The children to render
 * @param options Additional options
 * @returns The result of the render function.
 */
function testRender(
  state: Partial<RootState>,
  children: React.ReactNode,
  options?: TestRenderOptions,
): RenderResult & TestRenderResult {
  const store = mockStore(state);
  const theme = testTheme(options && options.width);
  const width = theme.breakpoints.values[
    options && options.width ? options.width : 'lg'];
  const globalMocks = mockGlobals(width);

  const renderResult = render(
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <MuiThemeProvider theme={theme}>
          <StylesProvider generateClassName={generateClassName}>
            {children}
          </StylesProvider>
        </MuiThemeProvider>
      </ConnectedRouter>
    </Provider>,
  );

  return {
    ...renderResult,
    globalMocks,
    store,
  };
}

export default testRender;
