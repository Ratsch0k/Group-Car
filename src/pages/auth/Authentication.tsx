import {
  Button,
  Dialog,
  DialogContent,
  DialogTitle,
  Divider,
  IconButton,
  makeStyles,
  createStyles,
  Container,
  Typography,
} from '@material-ui/core';
import ArrowBackIcon from '@material-ui/icons/ArrowBack';
import React, {useState, useEffect, useContext} from 'react';
import {
  Redirect,
  Route,
  Switch,
  useHistory,
  useRouteMatch,
  useLocation,
} from 'react-router-dom';
import {LoginForm} from './Login';
import {SignUpForm} from './SignUp';
import {useTranslation} from 'react-i18next';
import AuthContext from 'lib/context/auth/authContext';
import CloseIcon from '@material-ui/icons/Close';

type Theme = import('@material-ui/core').Theme;

const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    content: {
      paddingBottom: theme.spacing(3),
    },
    divider: {
      margin: theme.spacing(2),
    },
    dividerContainer: {
      position: 'relative',
    },
    dividerText: {
      paddingLeft: theme.spacing(0.5),
      paddingRight: theme.spacing(0.5),
      margin: 'auto',
      top: '50%',
      right: '50%',
      position: 'absolute',
      transform: 'translate(50%, -50%)',
      background: theme.palette.background.paper,
    },
    closeButton: {
      float: 'right',
    },
  }),
);

const Authentication: React.FC = () => {
  const match = useRouteMatch();
  const history = useHistory();
  const classes = useStyle();
  const {t} = useTranslation();
  const location = useLocation();
  const auth = useContext(AuthContext);

  const [loading, setLoading] = useState<boolean>(false);

  /**
   * Save the url from which the client navigated to this component
   */
  const [origin] = useState<string>('/');

  /**
   * Handles navigation to sign up page
   */
  const handleSignUp = () => {
    history.push(`${match.url}/sign-up`);
  };

  /**
   * Handles navigation to login page
   */
  const handleLogin = () => {
    history.push(`${match.url}/login`);
  };

  /**
   * Handles go back action
   */
  const handleBack = () => {
    history.goBack();
  };

  /**
   * Navigates to the page from which auth was called.
   */
  const onFinished = () => {
    history.replace(origin);
  };

  /**
   * Checks every time the `isLoggedIn` attribute changes
   * if the client is authenticated. If it is go to the origin.
   */
  useEffect(() => {
    if (auth.isLoggedIn) {
      onFinished();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [auth.isLoggedIn]);

  return (
    <Dialog open={true} fullWidth maxWidth='xs'>
      <DialogTitle>
        <IconButton
          data-testid='go-back'
          edge='start'
          disableRipple
          disabled={loading}
          onClick={handleBack}
          style={{visibility: match.isExact ? 'hidden' : 'visible'}}
        >
          <ArrowBackIcon/>
        </IconButton>
        {
          match.isExact ?
          t('auth.dialog.title.authenticate') :
          t('auth.dialog.title.' +
            location.pathname.replace(match.path + '/', ''))
        }
        <IconButton
          className={classes.closeButton}
          onClick={onFinished}
          disabled={loading}
          data-testid='close-dialog'
        >
          <CloseIcon />
        </IconButton>
      </DialogTitle>
      <DialogContent className={classes.content}>
        <Switch>
          <Route exact path={match.path}>
            <Button
              data-testid='to-login'
              fullWidth
              variant='outlined'
              color='primary'
              onClick={handleLogin}
            >
              {t('form.login')}
            </Button>
            <Container className={classes.dividerContainer}>
              <Divider orientation='horizontal' className={classes.divider}/>
              <Typography component='span' className={classes.dividerText}>
                {t('misc.or')}
              </Typography>
            </Container>
            <Button
              data-testid='to-signup'
              fullWidth
              variant='outlined'
              color='primary'
              onClick={handleSignUp}
            >
              {t('form.sign-up')}
            </Button>
          </Route>
          <Route path={`${match.path}/sign-up`}>
            <SignUpForm
              withSubmit
              onFinished={onFinished}
              setLoading={setLoading}
            />
          </Route>
          <Route path={`${match.path}/login`}>
            <LoginForm
              withSubmit
              onFinished={onFinished}
              setLoading={setLoading}
            />
          </Route>
          <Route path='*'>
            <Redirect to='/auth' />
          </Route>
        </Switch>
      </DialogContent>
    </Dialog>
  );
};

export default Authentication;
