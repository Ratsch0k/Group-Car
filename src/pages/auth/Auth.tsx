import React from 'react';
import {useHistory, useLocation} from 'react-router-dom';
import AuthenticationDialog from 'lib/components/Authentication';
import {AuthType} from 'lib/components/Authentication/AuthenticationDialog';

type TAuthType = import('lib/components/Authentication/AuthenticationDialog')
  .AuthType;

const Auth: React.FC = () => {
  const history = useHistory();
  const location = useLocation();

  // Extract if user wants to sign-up or login
  const allPaths = location.pathname.split('/');
  let lastPath: string | undefined =
    allPaths[allPaths.length - 1].split('?')[0];

  if (!(AuthType.includes(lastPath))) {
    lastPath = undefined;
  }

  const onAuthTypeChange = (type: TAuthType) => {
    if (type) {
      history.push(`/auth/${type}`);
    } else {
      history.replace('/auth');
    }
  };

  return (
    <AuthenticationDialog
      open={true}
      close={() => history.push('/')}
      initialAuthType={lastPath as TAuthType}
      onAuthTypeChange={onAuthTypeChange}
    />
  );
};

export default Auth;
