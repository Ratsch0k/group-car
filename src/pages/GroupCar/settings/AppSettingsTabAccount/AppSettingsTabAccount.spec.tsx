import "../../../../__test__/mockAxios";
import '../../../../__test__/mockI18n';
import {RootState} from "../../../../lib/redux/store";
import testRender from "../../../../__test__/testRender";
import AppSettingsTabAccount from "./AppSettingsTabAccount";
import {AuthState} from "../../../../lib/redux/slices/auth";

describe('AppSettingsTabAccount', () => {
  let state: Partial<RootState>;
  let resizeObserverMock;

  beforeEach(() => {
    state = {
      auth: {
        user: {
          id: 1,
          username: 'test',
          email: 'test@mail.com',
          createdAt: new Date('December 17, 2020 03:24:00').toUTCString(),
          updatedAt: new Date('December 17, 2020 03:24:00').toUTCString(),
          isBetaUser: false,
        }
      } as AuthState,
    }

    resizeObserverMock = jest.fn().mockImplementation(() => ({
      observe: jest.fn(),
      disconnect: jest.fn(),
    }));
    window.ResizeObserver = resizeObserverMock;

  });

  it('renders correctly', () => {
    let {baseElement} = testRender(
      state,
      <AppSettingsTabAccount index='/settings/account' value='/settings/account' />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});