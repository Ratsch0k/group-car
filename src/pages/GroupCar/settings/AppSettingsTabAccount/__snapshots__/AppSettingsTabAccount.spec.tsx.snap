// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSettingsTabAccount renders correctly 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root makeStyles-root MuiGrid-container MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root makeStyles-titleItem MuiGrid-item"
      >
        <h5
          class="MuiTypography-root makeStyles-title MuiTypography-h5"
        >
          <b>
            settings.account.title
          </b>
        </h5>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      />
    </div>
    <div
      class="MuiBox-root MuiBox-root"
    >
      <div
        class="MuiBox-root MuiBox-root makeStyles-root"
      >
        <div
          class="MuiGrid-root makeStyles-header MuiGrid-container MuiGrid-align-items-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <p
              class="MuiTypography-root makeStyles-title MuiTypography-body1"
            >
              <b>
                settings.account.userInfo
              </b>
            </p>
          </div>
          <div
            class="MuiGrid-root makeStyles-line MuiGrid-item"
          >
            <hr
              class="MuiDivider-root makeStyles-divider"
            />
          </div>
        </div>
        <div
          class="MuiAvatar-root MuiAvatar-circular makeStyles-avatar makeStyles-avatar"
        >
          <img
            alt="user-1"
            class="MuiAvatar-img"
            src="/api/user/1/profile-pic"
          />
        </div>
        <div
          class="MuiBox-root MuiBox-root makeStyles-values"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-wrap-xs-nowrap"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiGrid-root makeStyles-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-wrap-xs-nowrap"
              >
                <div
                  class="MuiGrid-root makeStyles-title MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-start"
                >
                  <p
                    class="MuiTypography-root makeStyles-label MuiTypography-body2"
                  >
                    misc.username
                  </p>
                </div>
                <div
                  class="MuiGrid-root makeStyles-content MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
                >
                  test
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiGrid-root makeStyles-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-wrap-xs-nowrap"
              >
                <div
                  class="MuiGrid-root makeStyles-title MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-start"
                >
                  <p
                    class="MuiTypography-root makeStyles-label MuiTypography-body2"
                  >
                    misc.email
                  </p>
                </div>
                <div
                  class="MuiGrid-root makeStyles-content MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
                >
                  test@mail.com
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="MuiGrid-root makeStyles-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-wrap-xs-nowrap"
              >
                <div
                  class="MuiGrid-root makeStyles-title MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-start"
                >
                  <p
                    class="MuiTypography-root makeStyles-label MuiTypography-body2"
                  >
                    misc.createdAt
                  </p>
                </div>
                <div
                  class="MuiGrid-root makeStyles-content MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
                >
                  12/17/2020, 3:24:00 AM
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root makeStyles-root"
      >
        <div
          class="MuiGrid-root makeStyles-header MuiGrid-container MuiGrid-align-items-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <p
              class="MuiTypography-root makeStyles-title MuiTypography-body1"
            >
              <b>
                settings.account.security.title
              </b>
            </p>
          </div>
          <div
            class="MuiGrid-root makeStyles-line MuiGrid-item"
          >
            <hr
              class="MuiDivider-root makeStyles-divider"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-wrap-xs-nowrap"
          style="overflow: hidden;"
        >
          <div
            class="MuiGrid-root makeStyles-title MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-start"
          >
            <p
              class="MuiTypography-root makeStyles-label MuiTypography-body2"
            >
              misc.password
            </p>
          </div>
          <div
            class="MuiGrid-root makeStyles-content MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
          >
            <div
              style="transition: height 250ms ease-in-out; width: 100%;"
            >
              <div>
                <div
                  class="makeStyles-root"
                >
                  <p
                    class="MuiTypography-root makeStyles-text MuiTypography-body1"
                  >
                    ********
                  </p>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button MuiButton-textPrimary"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      misc.edit
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
